light-client/src/light_client.rs

```diff
@@ -230,35 +232,32 @@ impl LightClient {
         }
     }
 
-    /// Look in the light store for a block from the given peer at the given height.
+    /// Look in the light store for a block from the given peer at the given height,
+    /// whatever its verification status.
+    ///
     /// If one cannot be found, fetch the block from the given peer.
     ///
     /// ## Postcondition
     /// - The provider of block that is returned matches the given peer.
-    #[post(ret.as_ref().map(|lb| lb.provider == self.peer).unwrap_or(true))]
+    #[post(ret.as_ref().map(|(lb, _)| lb.provider == self.peer).unwrap_or(true))]
     pub fn get_or_fetch_block(
         &self,
-        current_height: Height,
+        height: Height,
         state: &mut State,
-    ) -> Result<LightBlock, Error> {
-        let current_block = state
-            .light_store
-            .get(current_height, Status::Verified)
-            .or_else(|| state.light_store.get(current_height, Status::Unverified));
+    ) -> Result<(LightBlock, Status), Error> {
+        let block = state.light_store.get_non_failed(height);
 
-        if let Some(current_block) = current_block {
-            return Ok(current_block);
+        if let Some(block) = block {
+            return Ok(block);
         }
 
-        self.io
-            .fetch_light_block(self.peer, AtHeight::At(current_height))
-            .map(|current_block| {
-                state
-                    .light_store
-                    .insert(current_block.clone(), Status::Unverified);
+        let block = self
+            .io
+            .fetch_light_block(self.peer, AtHeight::At(height))
+            .map_err(ErrorKind::Io)?;
+
+        state.light_store.insert(block.clone(), Status::Unverified);
```

Comment:
It's strange to see mutations being done in a method that is called `get_or_fetch_block`. I know it has been here before but it indicates something being of about where we manage this concern. As now there is an implicit behavioural assumption that the fetched block will be in the light store going forward.

Reply:
Agreed it's not ideal. Allhough we don't really rely on this assumption, it's more of a performance optimization (ie. caching of fetched blocks). Perhaps the method should be renamed, or even refactored into a helper which caches blocks in the store and the original `get_or_fetch_block`. This would make the light store mutation more explicit at call site. Will think about it.

Reply:
Ideally behaviour like this (sounds like write-through caching) happens on a layer wrapping this component. Often referred to as decorators.This way each layer can be looked at in isolation and concerns are separated.

Reply:
I can give examples and/or go into more detail if it is of interest.

Reply:
> Ideally behaviour like this (sounds like write-through caching) happens on a layer wrapping this component. Often referred to as decorators.This way each layer can be looked at in isolation and concerns are separated.

Agreed, that's actually what I had in mind when I wrote "refactored into a helper which caches blocks in the store" but the name *decorator* didn't spring to my mind at the time :)

Reply:
Can we capture this somewhere so it doesn't get lost?

Reply:
Captured in #434.